cryptol-remote-api/python/cryptol/quoting.py

```diff
@@ -29,7 +29,9 @@ def to_cryptol_str(val : Union[CryptolValue, str, CryptolJSON]) -> str:
     elif isinstance(val, OpaqueValue):
         return str(val)
     elif isinstance(val, str):
-        return f'"{val}"'
+        # Sanitizing strings for Cryptol - see <string> of src/Cryptol/Parser/Lexer.x
+        val_san = val.replace('\\', '\\\\').replace('"', '\\"').replace('\n', '\\n')
```

Comment:
As far as I understand, the purpose of this function is to turn a Python value to the corresponding Cryptol value, written as a String.  Unfortunately, the notation for Cryptol string literals and Python ones are not exactly the same, so I believe the correct thing here would be to go through each character and use the relevant representation.  This would require careful examination of the Cryptol lexer, and would probably be a good 
opportunity to document the exact notation we are using.

Some differences that come to mind:
   * Make sure that all special escape sequences match
   * Numeric escape sequence are different (in Python "\10" is 8, because it uses octal, but in Cryptol it is 10, because we use base 10)
    * I think Python strings support Unicode, but Cryptol ones are just bytes, so we'd need to decide on what to do in that case---maybe just report an error, that this is not a valid Cryptol value.

Reply:
I agree w/ @yav that to get this right requires more care, and . Here is an idea I don't see has been discussed -- don't send strings to Cryptol at all, let Python handle strings (using `bytes`) and just send sequences of 8-bit numbers to Cryptol.

```
list(bytes("hello, \10", 'ascii'))
[104, 101, 108, 108, 111, 44, 32, 8]
```

Of course, if we did this we would want to send the type along, so that Cryptol could infer the sequence had 8-bit values -- "[104, 101, 108, 108, 111, 44, 32, 8] : [8][8]".

Reply:
What about cases where Cryptol and Python disagree? I am leaning towards Python, since Python is the source language here.

```
Cryptol> :s base=10
Cryptol> "hello, \&"
[104, 101, 108, 108, 111, 44, 32]
```
```
>>> list(bytes("hello, \&", 'ascii'))
[104, 101, 108, 108, 111, 44, 32, 92, 38]
```

Reply:
These are excellent points @yav and @weaversa – thanks for the comments!

> This would require careful examination of the Cryptol lexer, and would probably be a good
opportunity to document the exact notation we are using.

I took a quick look, and it appears that the strategy is to lex strings using Haskell's `read` function! By that I mean the strategy is to collect all the characters in the string (including the quote delimiters and with no processing of escape sequences), then use the built-in Haskell [`reads` function](https://github.com/GaloisInc/cryptol/blob/57226efb27e801dc911b8caf865bd3d57425fc1c/src/Cryptol/Parser/LexerUtils.hs#L109).

So instead of trying to capture the complexity of Haskell strings in this tiny bit of Python code, I really like @weaversa's solution. It's simple and leaves no room for ambiguity.

As per your second comment, Sean, I agree that we should send over the bytes corresponding to the string from Python's perspective. The alternative – trying to figure out what the bytes would be if the literal were instead parsed as Cryptol – feels about as difficult as just figuring out how to properly send the string literal.